File: question/leetcode/contest/weekly-contest-419/q4.cpp
#include <bits/stdc++.h>

using namespace std;

typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef vector<vector<vector<int>>> vvvi;

typedef pair<int, int> pii;
typedef vector<pair<int,int>> vii;
typedef vector<vector<pair<int,int>>> vvii;
typedef vector<tuple<int, int, int>> viii;

typedef tuple<int, int, int> tiii;

typedef long long ll;

typedef vector<ll> vl;
typedef vector<vector<ll>> vvl;
typedef vector<vector<vector<ll>>> vvvl;

typedef pair<ll, ll> pll;
typedef vector<pair<ll,ll>> vll;
typedef vector<vector<pair<ll,ll>>> vvll;
typedef vector<tuple<ll, ll, ll>> vlll;

typedef tuple<ll, ll, ll> tlll;

#define f first
#define s second
#define pb push_back
#define mp make_pair
#define range(i,b) for (ll i = 0; i < b; i++)
#define rep(i,a,b) for (ll i = a; i < b; i++)
#define repr(i,a,b) for (ll i = a; i > b; i--)
#define sq(a) (a)*(a)
#define all(a) a.begin(), a.end()
#define rall(a) a.rbegin(), a.rend()
#define ceild(a,b) ((a)%(b) ? ((a)/(b) + 1) : (a)/(b))
#define TxtIO freopen("input.txt","r",stdin); freopen("output.txt","w",stdout)

const ll mod = 1e9 + 7;

const int au = 65;
const int al = 97;
const int zu = 90;
const int zl = 122;

const int lt = 32;

class Solution {
public:
    vector<int> findXSum(vector<int>& v, int k, int x) {
        int n = v.size();

        vector<pair<int,int>> c;
        unordered_map<int, int> m;
        int l = 0;
        int t = 0;

        vi a;

        rep(r,0,n){
            c[v[r]]++;
            
            if(r>=k-1){

                vii count;
                for(auto [k,v]: c){
                    count.pb({v,k});
                }

                sort(rall(count));
                int t = 0;
                int s = min(int(count.size()),x);
                rep(i,0,s){
                    t+= count[i].f*count[i].s;
                }
                a.pb(t);
                
                c[v[l]]--;
                l++;
            }
        }
        return a;
    }
};

File: question/leetcode/contest/weekly-contest-419/q4.py
from sortedcontainers import SortedList

class Solution:
    def findXSum(self, v: List[int], k: int, x: int) -> List[int]:
        n = len(v)

        s = SortedList()
        d = defaultdict(int)
        l = 0
        t = 0
        a = []

        for r in range(len(v)):
            # print(r)
            # print(s)
            # print(t)

            b = v[r]
            size = d[b]
            d[b]+=1

            i = s.bisect_left((-size,-b))
            if i<len(s) and s[i] == ((-size,-b)):
                s.discard((-size,-b))

                if(i<=x-1):
                    t-=size*b

            s.add((-(size+1),-b))
            j = s.bisect_left((-size-1,-b))
            if(j<=x-1):
                t+=(size+1)*b

            if(j<=x-1 and not i<=x-1):
                t-=-s[x][0]*-s[x][1]


            # print(s)
            # print(t)

            if r>k-1:
                b = v[l]
                size = d[b]
                d[b]-=1

                i = s.bisect_left((-size,-b))
                if i<len(s) and  s[i] == ((-size,-b)):
                    s.discard((-size,-b))

                    if(i<=x-1):
                        t-=size*b

                s.add((-(size-1),-b))
                j = s.bisect_left((-size+1,-b))
                if(j<=x-1):
                    t+=(size-1)*b

                if(not j<=x-1 and i<=x-1):
                    t+=-s[x-1][0]*-s[x-1][1]

                l+=1

            # print(s)
            # print(t)

            if r>=k-1:
                a.append(t)

        return a